### PR TITLE
Fix SIGINT/SIGTERM [V5]

### PR DESCRIPTION
**Pull request by apahim**
v5:
- Add docs.
- Selftests:
  - Use `proc.stdout.read()` instead of `proc.communicate()`
  - Improve comments.

v4: #2150 
- Selftests:
  - Refactor to reuse the common methods.
  - Use `stderr=subprocess.STDOUT` 

v3: #2140 
- Change the approach to fix this issue. Previously, I was trying to handle the signals from the main process all the way to the test sub-process, through all the layers in between. That was proven hand to control and with a big number of corner cases. Now, the approach is rather simple:
  - When a SIGINT in received by the main process, propagate the SIGINT for all the children.
  - When a SIGTERM is received by the main process, send a SIGKILL for all the children.
- Tests where rewritten to cover all the scenarios we have so far.

v2: #2137 
- Fix message on SIGINT (new commit)
- Improve SIGINT selftest.
- New commit to fix SIGINT/SIGTERM behavior on avocado.runner and process.Subprocess. See commit message for details.

v1: #2130

### CODE REVIEW

**Comment by apahim** [comment]
[BUG: handling of SIGINT on test runner freezes it](https://trello.com/c/5RPZPH4F/1068-bug-handling-of-sigint-on-test-runner-freezes-it)

File: docs/source/ReferenceGuide.rst [hunk 1]
```diff
@@ -653,4 +653,31 @@ be a valid one:
...
+
+Avocado normal operation is related to run code written by
+users/test-writers. It means the test code can carry its own handlers
+for different signals or even ignore then. Still, as the code is being
+executed by Avocado, we have to make sure we will finish all the
+subprocesses we create before ending our execution.
+
+Signals sent to the Avocado main process will be handled as follows:
+
+- SIGSTOP/Ctrl+Z: On SIGSTOP, Avocado will pause the execution of the
```
Comment (ldoktor) [inline docs/source/ReferenceGuide.rst]: This is suppose to be SIGTSTP. SIGSTOP can't be caught.

Comment (apahim) [inline docs/source/ReferenceGuide.rst]: Ops...

File: docs/source/ReferenceGuide.rst [hunk 2]
```diff
@@ -653,4 +653,31 @@ be a valid one:
...
+  subprocesses, while the main process will still be running, waiting
+  for the subprocesses to finish. A new SIGSTOP will make the
```
Comment (ldoktor) [inline docs/source/ReferenceGuide.rst]: timeout timer still running...

File: docs/source/ReferenceGuide.rst [hunk 3]
```diff
@@ -653,4 +653,31 @@ be a valid one:
...
+  subprocesses to resume the execution.
+- SIGINT/Ctrl+C: This signal will be forwarded to the test process and
+  Avocado will wait until it's finished. If the test process does not
+  finish after receiving a SIGINT, user can send a second SIGINT (after
+  the 2 seconds ignore period). The second SIGINT will make Avocado
+  to send a SIGKILL to the whole subprocess tree and then complete the
+  main process execution.
+- SIGTERM: This signal will make Avocado to terminate immediately. A
+  SIGKILL will be sent to the whole subprocess tree and the main process
+  will exit without completing the execution.
```
Comment (ldoktor) [inline docs/source/ReferenceGuide.rst]: I'd add a note that we do best-effort approach, therefor in case of fork-bomb some of the newly created processes might not be killed.

File: selftests/functional/test_interrupt.py [hunk 4]
```diff
@@ -134,59 +166,120 @@ def test_well_behaved(self):
...
+        if not wait.wait_for(self._no_test_in_process_table, timeout=2):
+            self.fail('Avocado left processes behind.')
+
+        # Make sure the Interrupted test sentence is there
+        self.assertIn('Terminated\n', self.proc.stdout.read())
 
     def tearDown(self):
         shutil.rmtree(self.tmpdir)
 
+    def _has_children(self):
```
Comment (ldoktor) [inline selftests/functional/test_interrupt.py]: It's generally better to include function before their usage.

File: selftests/functional/test_interrupt.py [hunk 5]
```diff
@@ -134,59 +166,120 @@ def test_well_behaved(self):
...
+        return len(psutil.Process(self.proc.pid).children()) > 0
```
Comment (ldoktor) [inline selftests/functional/test_interrupt.py]: Also you do spawn 2 test processes, I think it'd be better to wait for at least `> 1` (to avoid proceeding when sysinfo is being generated). It'd also fail in case of daemons, but it should behave way better (alternative would be to create a file inside the test, but IMO that's not necessary)

File: selftests/functional/test_interrupt.py [hunk 6]
```diff
@@ -67,64 +77,86 @@ def test_badly_behaved(self):
...
-                            bad_test_processes.append(p_obj)
-                # psutil.NoSuchProcess happens when the original
-                # process already ended and left the process table
-                except psutil.NoSuchProcess:
-                    pass
-
-            return len(bad_test_processes) == 0
-
-        wait.wait_for(wait_until_no_badtest, timeout=2)
+        if not wait.wait_for(self._no_test_in_process_table, timeout=2):
```
Comment (ldoktor) [inline selftests/functional/test_interrupt.py]: nitpick: Could be `self.assertTrue` but I'm fine with this...

File: selftests/functional/test_interrupt.py [hunk 7]
```diff
@@ -67,64 +77,86 @@ def test_badly_behaved(self):
...
+            self.fail('Avocado left processes behind.')
+
+        output = self.proc.stdout.read()
```
Comment (ldoktor) [inline selftests/functional/test_interrupt.py]: The `communicate` is the recommended way of obtaining the output and finish the communication with the process, can I know why did you change it to `stdout.read`?

File: selftests/functional/test_interrupt.py [hunk 8]
```diff
@@ -67,64 +77,86 @@ def test_badly_behaved(self):
...
+        # Make sure the Interrupted requested sentence is there
+        self.assertIn('Interrupt requested. Waiting 2 seconds for test to '
+                      'finish (ignoring new Ctrl+C until then)', output)
         # Make sure the Killing test subprocess message did appear
-        self.assertIn('Killing test subprocess', proc.get_output())
+        self.assertIn('Killing test subprocess', output)
```
Comment (ldoktor) [inline selftests/functional/test_interrupt.py]: To be honest I imagined even greater code-reuse, but this at least shares the most delicate parts so I'm fine with this way at this point.

**Comment by clebergnu** [comment]
What do you mean by "finish the communication"?

**Comment by clebergnu** [comment]
Now I see, you mean that it waits for the process to finish.

File: selftests/functional/test_interrupt.py (hunk 4 above)
Comment (apahim) [inline selftests/functional/test_interrupt.py]: It makes no difference for class methods.

File: selftests/functional/test_interrupt.py (hunk 6 above)
Comment (apahim) [inline selftests/functional/test_interrupt.py]: Sure, I'm ok with both.

File: selftests/functional/test_interrupt.py (hunk 8 above)
Comment (apahim) [inline selftests/functional/test_interrupt.py]: Yes, I considered what would be the best balance between readability and code re-use. Since this is a selftest (and you're fine), I think we are good here. More code re-use would make it a little harder to follow, imo.

File: selftests/functional/test_interrupt.py (hunk 7 above)
Comment (apahim) [inline selftests/functional/test_interrupt.py]: I don't see a reason why you couldn't :)
I changed because I was requested to && at this point we already went through `self._is_finished || process.kill_process_tree(self.proc.pid)`, so the `communicate()` added value in waiting for the process to finish, in this case, is not really important.

File: selftests/functional/test_interrupt.py (hunk 5 above)
Comment (apahim) [inline selftests/functional/test_interrupt.py]: Ok

Comment (apahim) [inline selftests/functional/test_interrupt.py]: Actually, this is checking children. So, it's the number of subprocesses, excluding the main process. It means we are good with `> 0`.

**Comment by apahim** [comment]
@ldoktor @clebergnu, waiting for an agreement between the two of you regarding the `proc.stdout.read()` vs `proc.communicate()`. I ok with both.

**Comment by clebergnu** [comment]
@ldoktor you say that "communicate is the recommended way of obtaining the output and finish the communication with the process", and I agree.  @apahim explained that the checking the process termination is already done.  I proposed using `proc.stdout.read()` to improve code readability (its meaning is much more straightforward IMO).

@ldoktor do you see any problem with the current approach?  If not, @apahim can go ahead and send a v6 without reverting that chunk.

**Comment by ldoktor** [comment]
I'm fine with both with slight preference towards `communicate`. So just the documentation and such and we're ready to ship.

**Comment by apahim** [comment]
Thanks.
